### PR TITLE
Footsides

### PR DESCRIPTION
**Pull request by jo-pol**
@Veronika

The image on the left shows an example of `target/test/fringes/` resulting from the new `Fringes` class and the corresponding test class `FringesDemo`. The red links are supposed to be the pairs required to start the lace. The green links are pairs coming out of the to be developed footside. The semi-transparent dots are nodes requiring a pair going into the footside, darker dots require two pairs.

The image on the right shows the result of the <strike>to be implemented</strike> FootSide [property](https://github.com/d-bl/GroundForge/pull/67/files?diff=unified#diff-a49deea019654e985d1e27c7be177bafR142) of the Fringes class.

diagrams
[pattern](https://d-bl.github.io/GroundForge/?tiles=bricks&matrix=5-K-5-K-_-L-O-L-O_5-L---H-_-E-H-E-H&rows=22&cols=23&left=0&up=0&stitches=)

The right hand side graph would allow thread diagrams for all configurations, but the footsides are rubbish. Where a link has to skip a previous green link to find a free node, split both and connect at an additional node as shown step by step below. That can still leave us with parallel links, their source and target nodes <strike>should be</strike> are merged.

naamloos
[pattern](https://d-bl.github.io/GroundForge/?tiles=bricks&matrix=5-E-H-%0D%0A-5-5--%0D%0AL-5-O-&stitches=&rows=22&cols=23&left=0&up=0)

Create PNG's from the generated SVG's with InkScape and bash:

    for i in *.svg; do 'path/to/inkscape.exe' $i --export-png=`echo $i | sed -e 's/svg$/png/'`; done
[pattern]: https://d-bl.github.io/GroundForge/?tiles=bricks&matrix=5-E-H-%0D%0A-5-5--%0D%0AL-5-O-&stitches=&rows=22&cols=23&left=0&up=0&transparency=0&stitches=

### CODE REVIEW

File: src/main/scala/dibl/Fringes.scala [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    def createLinks(cols: Inclusive): IndexedSeq[Link] = {
+      for {
+        row <- bottomTargetRow to (topTargetRow, -1)
+        col <- cols
+        links = popLinks(row, col)
+        _ = pushLinks(row, col)
+      } yield links
+    }.flatten
+
+    createLinks(leftTargetCol to rightTargetCol) ++ createLinks (rightTargetCol to (rightTargetCol - 1, -1))
```
Comment (jo-pol) [inline src/main/scala/dibl/Fringes.scala]: Thanks @rvanHeest for your explanation of the shunting-yard algorithm, still have to test it.

Comment (rvanheest) [inline src/main/scala/dibl/Fringes.scala]: Looks great! You're welcome. Happy to have helped you out. Also great to see you using all these for-comprehensions! However, don't over-use them either. For example, in both `draw` functions below a simple `map` might work just as fine.

**Question:** Is there a specific reason for this `flatten`? I mean, if you change the `=` in `links = popLinks(row, col)` into a `<-`, you probably don't need the `flatten`. Think about the `=` as a `map` and `<-` as a `flatMap` in this case...

**Comment by jo-pol** [comment]
Both push and pop actions are about zero to two elements. I need to do both pops before pushing anything.  I shouldn't call the push action twice or not at all depending on the number of pops. It might be clearer to put both actions in the yield block.

File: src/main/scala/dibl/Fringes.scala (hunk 1 above)
Comment (jo-pol) [inline src/main/scala/dibl/Fringes.scala]: The answer by mail landed on the main page of the pull request what makes the discussion hard to follow.

You are right about the draw functions.

Rephrasing the answer to your question: `popLinks` returns another type (nested tuples) and can't be mixed with the two previous expressions (plain integers). Moreover, changing the `=` in `<-` would change the frequency `pushLinks` gets called, it should be called once, regardless of the number of items return by `popLinks`.

Just tested it, made a mistake in the first call to `createLinks` and need to clear the stack between the calls. Otherwise :tada:

**Comment by jo-pol** [comment]
The algorithm occasionally mixes up the left and right pair. The arrow in the image indicate where the pairs don't cross as a result. The pictured [pattern](LOCAL-PATH-TO/docs/index.html?tiles=bricks&matrix=586-+-789+2111+-4-4&color1=FF1A05&color2=FF1F05&color3=1AFF0F&color4=2DFF17&color5=0A0CFF&color6=0A0CFF&color7=FF1ACA&color8=FF0F9B&color9=82FFD1&color10=82FFD1&color11=FFDA82&color12=FFDA82&color13=A385FF&color14=A385FF&color15=13871C&color16=13871C&color=000000&footside=ttctc&stitches=&rows=23&cols=14&left=1&up=1&transparency=0)

naamloos

File: src/main/scala/dibl/PairDiagram.scala [hunk 2]
```diff
@@ -15,143 +15,130 @@
...
-      assignPairNrs(nodes, links)
-      PairDiagram(nodes, links)
+  def apply(triedSettings: Try[Settings]): PairDiagram = if (triedSettings.isFailure)
+    PairDiagram(Seq(Props("title" -> triedSettings.failed.get.getMessage, "bobbin" -> true)), Seq[Props]())
+  else {
+    val settings: Settings = triedSettings.get
+    val fringes = new Fringes(triedSettings.get.absM)
+    val sources: Seq[Cell] = fringes.newPairs.map { case (source, _) => source }
+    val plainLinks: Seq[Link] = fringes.newPairs ++ fringes.leftFootSides ++ fringes.coreLinks ++ fringes.rightFootSides
+    val linksByTarget: Map[Cell,Seq[Link]] = replaceYsWithVs(plainLinks.groupBy { case (_, target) => target })
```
Comment (jo-pol) [inline src/main/scala/dibl/PairDiagram.scala]: To generate a proper thread diagram out of a pair diagram, the links should be listed clock wise per target. Not 100% sure, but it looks like that `groupBy` keeps original relative positions of the links.

The diagram below illustrates an example of `plainLinks` which has problems with the order. The new pairs are red, footsides in purple and core links grey.
naamloos
For the node in the top right corner, the red link will come before the grey one, what is not clockwise. On the left we occasionally see two purple links arriving at the same target, the long one comes before the short, that is also not clock wise.

Performance wise it might be better to reuse `absM` and replace the original links with the ones altered by the `Fringes class`.
